### DISCUSSION TITLE
Install MeshAgent in WinPE?

### DISCUSSION THREAD

**Discussion by shinomen** [Ylianst/MeshCentral#6348] [General]
I was wondering if anyone knew of a way where we could have the MeshAgent installed under WinPe?  We utilize different boot utilities from USB as well as from PXE boot over the network.  I was wondering if there was a way where we could inject a MeshCentral Agent into a PXE or WIM image we have so that when that environment is started we could gain a connection to the computer remotely.
For example, when I'm working remotely I can get a local user to PXE boot from the network where our disk cloning environment will boot up.  It would be nice if that environment had the MeshCentral Agent running when it's started so that it can be visible from our MeshCentral controller and we could then connect to the WinPE environment that is running so that we can continue the disk cloning process.

**Comment by si458** [comment]
So long ad it's windows I don't see why it wouldn't work?
Just include the exe in your image, then maybe get ur image to auto run `meshagent.exe -fullinstall` to install the agent into program files OR `meshagent.exe connect` to connect and keep it running until the app closes?

**Comment by PathfinderNetworks** [comment]
That's a great idea.  I'm going to give it a go once I have some time to try it.  There are many WinPE boot utilities we use as well and, having MeshCentral, would be a nice option as we could have an onsite tech boot the device with the WinPE image and then a more senior remote tech could take over (for example). 
Here is the Microsoft help guide for how to customize WinPE images.  There is a section for adding apps. 
https://learn.microsoft.com/en-us/windows-hardware/manufacture/desktop/winpe-mount-and-customize?view=windows-11

**Comment by shinomen** [comment]
Well, I can report back that it works pretty well!  I added the single MeshAgent.exe file to the wim image that I mounted and edited the startnet.cmd file under the windows\system32 folder to have the meshagent.exe -full install like you mentioned and committed the changes to the wim.  I PXE booted and sure enough the machine showed up under MeshCentral and I was able to connect right away!
Thanks!

**Reply by si458** [reply]
Now that truly is amazing! 
Random through.
If u restarted and rebooted and reloaded over network, does it connect back to the same machine, or does a new machine appear?
also what about if multiple machines pxe boot the same image?
Does separate machines all appear or is it then same machine?

**Reply by shinomen** [reply]
I tried that.   When rebooting, the machine gets a new name (hyper-v virtual machine) and shows up as a separate computer under the site.  I would imagine that everytime it reboots, it will continue to generate a new name.  
Additionally, if I remove the device from the group while the winpe is still running, it will reconnect and create the machine but then the remote access does not work until the winpe machine is restarted.  
I was able to access files option and I could send messages, but other things like chat, remove vnc, remote rdp, and actions like restart, and shutdown didn't work.  
For me this is excellent though because I can remotely access the machine.  I had to do a clone yesterday where I was essentially babysitting the machine resizing the hard drive for 2 hours on-site.  If I had this tool, I could have started the process and left the client but still have been able to remotely check in on this machine.
It really was not hard to implement and will be such a valuable addition. :-)

**Reply by si458** [reply]
ok so one thing to check is make sure each machine has a unique UUID (identifier) for its motherboard + bios + windows os !
as the meshagent uses this mixed with mac addresses to make each device unique
so if one thing is the same across multiple computers (or virtual ones) this can mix the agents up by mistake!

**Reply by si458** [reply]
also maybe the `connect` option rather then `-fullinstall` might be better
in your winpe environment as its just installing an image on the remote device

then you could for example run `-fullinstall` inside the OS image instead to have it always use the same nodeids etc across reboots!

just an amazing idea! i dont really use pxe, at most i use netboot.xyz to install an os if i dont have it handy on a usb drive haha

**Reply by shinomen** [reply]
I was able to play some more with this and using the connect vs -fullinstall does work better.  The device goes away after some time when the remote winpe is shut off (maybe about 2 min).  That's excellent because I won't have to clean up that group periodically. 
I don't always use PXE but it's nice to have it if I'm not on-site.  
Typically I use a USB bootable (Ventoy) to start up computers to do cloning or partition resizing.  Again, I'm going to put this meshagent inside the WinPE environment wim files on that USB boot utility too so that I can go away and still access it remotely if needed.
I'm a one man MSP so the more I can do remotely without having to be tied down, the more I can accomplish.

Thanks again!

**Reply by si458** [reply]
glad it works!
in theory the is another option called `run` with does a `console agent` connect
the different being listed below
```
Mesh Agent available switches:
  run               Start as a console agent.
  connect           Start as a temporary console agent.
  start             Start the service.
  restart           Restart the service.
  stop              Stop the service.
  state             Display the running state of the service.
  -signcheck        Perform self - check.
  -install          Install the service from this location.
  -uninstall        Remove the service from this location.
  -nodeid           Return the current agent identifier.
  -info             Return agent version information.
  -resetnodeid      Reset the NodeID next time the service is started.
  -fulluninstall    Stop agent and clean up the program files location.
  -fullinstall      Copy agent into program files, install and launch.

                    The following switches can be specified after -fullinstall:

     --WebProxy="http://proxyhost:port"      Specify an HTTPS proxy.
     --agentName="alternate name"            Specify an alternate name to be provided by the agent.
```
and in theory the `run` is meant to be the same as running `meshagent.exe` without anything
but from my testing, it seems to do the same thing as `connect` and delete the agent afterwards,
so who knows haha

**Reply by shinomen** [reply]
So it appears that using just the 'connect' switch causes the device to lose remote desktop connection abilities if the network connection is disrupted.  I manually ran the meshagent under windows pe using the 'run' command and it was able to maintain the ability to remote desktop to the machine even with network connection disruptions.  
With 'connect' it was still seen in the meshcentral dashboard so the connectivity was there, but the remote feature disappeared when the connection on the windows pe machine was disrupted then restored.

**Reply by si458** [reply]
thanks for testing and verifying! so indeed the descriptions really are correct
`run` run all the time and if internet drops reconnect
`connect` run all the time BUT if internet drops, never reconnect!
so in a sense `temporary` for the `connect` is correct! nice!
